apps/ai-image-generator/backend/src/upload-images.ts

```diff
@@ -0,0 +1,54 @@
+import { PlainClientAPI, UploadProps } from 'contentful-management';
+import { ImageWithStream, ImageWithUpload } from './types';
+
+const UPLOAD_DOMAIN: Record<string, URL> = {
+  us: new URL('https://s3.us-east-1.amazonaws.com/upload-api.contentful.com'),
+  eu: new URL('https://s3.us-east-1.amazonaws.com/upload-api.contentful.com'),
+};
+
+export class SharpStreamsToUrl {
+  constructor(
+    readonly imagesWithStreams: ImageWithStream[],
+    readonly cmaClient: PlainClientAPI,
+    readonly spaceId: string
+  ) {}
+
+  async execute(): Promise<ImageWithUpload[]> {
+    return Promise.all(
+      this.imagesWithStreams.map((imageWithStream) => this.streamToImage(imageWithStream))
+    );
+  }
+
+  private async streamToImage(imageWithStream: ImageWithStream): Promise<ImageWithUpload> {
+    const uploadProps = await this.createUpload(imageWithStream);
+    const url = this.urlFromUpload(uploadProps);
+    const upload = { ...uploadProps, url };
+    return {
+      url: imageWithStream.url,
+      imageType: imageWithStream.imageType,
+      upload,
```

Comment:
I think I might be getting confused here, so is this url object the url pointing to the Upload URL? and not DALL-E (my original understanding was the URL property would be the DALL-E one for backwards compatibility)? Ultimately it would make sense to me that the backend supplies the upload URL so we don't need to configure them on the frontend _if you are already doing the logic here_ (goes back to my original question in the thread I think so apologies for repetition here!) I believe the only information we need on the frontend to render and save the image is the `url` and the `imageType` (png etc)

Reply:
okay wait I think I am piecing this together - I do see the url being sent as part of the upload object. So I am wondering if all we need is the URL then?

Reply:
Yeah I think for now, we can keep the entire payload, however in the end I think it might make most sense to condense this to the shortest amount of info the client needs